lib/src/coap_client.dart

```diff
@@ -35,338 +55,353 @@ class CoapClient {
-        as CoapRequest?;
-    return send(accept(
-        request!.setPayloadMediaRaw(payload, format) as CoapRequest,
-        acceptVal));
-  }
-
-  /// If match
-  Future<CoapResponse> putIfMatch(
-      String payload, int format, List<typed.Uint8Buffer> etags) {
-    request ??=
-        CoapRequest.newPut().setPayloadMedia(payload, format) as CoapRequest?;
-    return send(ifMatch(
-        request!.setPayloadMedia(payload, format) as CoapRequest, etags));
-  }
-
-  /// If match byte payload
-  Future<CoapResponse> putIfMatchBytePayload(
-      typed.Uint8Buffer payload, int format, List<typed.Uint8Buffer> etags) {
-    request ??= CoapRequest.newPut().setPayloadMediaRaw(payload, format)
-        as CoapRequest?;
-    return send(ifMatch(
-        request!.setPayloadMediaRaw(payload, format) as CoapRequest, etags));
-  }
-
-  /// If none match
-  Future<CoapResponse> putIfNoneMatch(
-      String payload, int format, List<typed.Uint8Buffer> etags) {
-    request ??=
-        CoapRequest.newPut().setPayloadMedia(payload, format) as CoapRequest?;
-    return send(ifNoneMatch(
-        request!.setPayloadMedia(payload, format) as CoapRequest, etags));
-  }
-
-  /// If none match byte payload
-  Future<CoapResponse> putIfNoneMatchBytePayload(
-      typed.Uint8Buffer payload, int format, List<typed.Uint8Buffer> etags) {
-    request ??= CoapRequest.newPut().setPayloadMediaRaw(payload, format)
-        as CoapRequest?;
-    return send(ifNoneMatch(
-        request!.setPayloadMediaRaw(payload, format) as CoapRequest, etags));
-  }
-
-  /// Delete
-  Future<CoapResponse> delete() {
-    request ??= CoapRequest.newDelete();
-    return send(request!);
-  }
-
-  /// Validate
-  Future<CoapResponse> validate(List<typed.Uint8Buffer> etags) {
-    request ??= CoapRequest.newGet();
-    return send(eTags(request!, etags));
+  Future<bool> ping({Duration? timeout}) async {
+    final request = CoapRequest(CoapCode.empty, confirmable: true);
+    request.token = CoapConstants.emptyToken;
+    await _prepare(request);
+    _endpoint!.sendEpRequest(request);
+    await _waitForReject(request, timeout ?? this.timeout);
+    return request.isRejected;
+  }
+
+  /// Sends a GET request.
+  /// block2Size: early negotiation blocksize (16, 32, 64, 128, 256, 512, or 1024)
+  Future<CoapResponse> get(
+    String path, {
+    int accept = CoapMediaType.textPlain,
+    int type = CoapMessageType.con,
+    List<CoapOption>? options,
+    int block2Size = 0,
+    int maxRetransmit = 0,
+    Duration? timeout,
+    CoapMulticastResponseHandler? onMulticastResponse,
+  }) {
+    final request = CoapRequest.newGet();
+    _build(request, path, accept, type, options, block2Size, maxRetransmit);
+    return send(request,
+        timeout: timeout, onMulticastResponse: onMulticastResponse);
+  }
+
+  /// Sends a POST request.
+  /// block2Size: early negotiation blocksize (16, 32, 64, 128, 256, 512, or 1024)
```

Comment:
Yeah, I've seen enum used in other libs for SZX, just like it is specified in the RFC. Makes it more transparent, now the payload size will be rounded and unless the user has paid attention to the doc comment they might be a bit surprised. I can change to an enum and have a look at a block1 setting as well.

Reply:
Also I think all block1 requests use early negotiation always (since the first block will contain data anyway), but I could be wrong: https://datatracker.ietf.org/doc/html/rfc7959

Reply:
> I suggest holding off until we have enhanced enums in Dart 2.17, created issue: https://github.com/shamblett/coap/issues/80

That sounds reasonable :)

> Also I think all block1 requests use early negotiation always (since the first block will contain data anyway), but I could be wrong: https://datatracker.ietf.org/doc/html/rfc7959

Hmm, I think by indicating the block1 value you could set the client's preferred/initial block size (c. f. figure 9). Having another look into the source code, though, I am not sure if "manually" adding a block1 option/setting the size would actually be processed at the moment. Maybe this is another potential issue that can be addressed in a follow-up PR :)

Reply:
Yeah, but what makes the negotiation different is that for every block in a block1 request we set:
```dart
    block
        .addOption(CoapBlockOption.fromParts(optionTypeBlock1, num, szx, m: m));
```

This includes the first request/block. There will still be a negotiation, where the block1 size in subsequent messages is adjusted to the smallest of the client's preferred block1 size and the server's maximum block1 size.

In a block2 get for example, the get request itself is not blockwise, so we can negotiate early by including the block2 options as an instruction for the blockwise response from the server.

Reply:
But in general I'm wondering if we should include this option to alter block size for specific requests at all, I just kept it because it was exposed to the user in the first place.

The block size is also set depending on connectivity (likelihood of IP packet fragmentation and such), so it would make sense to use the same settings for the whole client since it now maps 1:1 with a specific server, and this is specified in the config already. If you want to push larger payloads to the client you wait longer, you typically don't increase the block size.

Reply:
Example: https://github.com/shamblett/coap/pull/76/commits/8dc6fb302e08398a7324e25ab3966f75806e0d66

So there is no `defaultBlockSize` but a `preferredBlockSize` as per the RFC. This is what the client as a whole prefers. Then you can choose whether or not to instruct the server to use the client's preference in its block2 response.

The whole negotiation is "piggybacked" with the requests/responses, so for block1 you can't negotiate early unless you want an extra roundtrip (very expensive), because the first block1 request already contains a payload.

Reply:
Oh, that is a very elegant solution :) Very cool, thank you! :)

Reply:
Thank you, too, for focusing on the uglier parts of my refactor to make it better :)